comms/core/src/connection_manager/dialer.rs

```diff
@@ -511,87 +545,121 @@ where
     /// Returns ownership of the given `DialState` and a success or failure result for the dial,
     /// or None if the dial was cancelled inflight
     async fn dial_peer(
-        dial_state: DialState,
+        mut dial_state: DialState,
         noise_config: &NoiseConfig,
         transport: &TTransport,
         network_byte: u8,
     ) -> (
         DialState,
         Result<(NoiseSocket<TTransport::Output>, Multiaddr), ConnectionManagerError>,
     ) {
-        let mut addr_iter = dial_state.peer().addresses.iter();
+        let addresses = dial_state.peer().addresses.clone().into_vec();
         let cancel_signal = dial_state.get_cancel_signal();
-        loop {
-            let result = match addr_iter.next() {
-                Some(address) => {
-                    debug!(
-                        target: LOG_TARGET,
-                        "Attempting address '{}' for peer '{}'",
-                        address,
-                        dial_state.peer().node_id.short_str()
-                    );
+        for address in addresses {
```

Comment:
Yeah, I did consider that. In this case tokio would be better than rayon, because it would wait on the network dial. Rayon would be a better choice if it were CPU bound. The dials already, for a few cases, are parallelized during a call to each peer, but you are correct in saying that if a peer had many addresses that it would take a while.

Some things to note though:
1. The addresses for a peer are always sorted by their quality score, so the best addresses will be at the front.
2. Addresses that have failed previously are moved to the back of the list, so will be tried after all others.

So from these, it makes sense to keep it trying addresses in synch and to exit on the first dialled address